src/IpAddress.php

```diff
@@ -175,54 +175,102 @@ public function __invoke(ServerRequestInterface $request, ResponseInterface $res
      * @param  ServerRequestInterface $request PSR-7 Request
      * @return string
      */
-    protected function determineClientIpAddress($request)
+    protected function determineClientIpAddress($request): ?string
     {
-        $ipAddress = '';
+        $ipAddress =  null;
 
         $serverParams = $request->getServerParams();
         if (isset($serverParams['REMOTE_ADDR'])) {
             $remoteAddr = $this->extractIpAddress($serverParams['REMOTE_ADDR']);
-            if ($this->isValidIpAddress($remoteAddr)) {
+            if (filter_var($remoteAddr, FILTER_VALIDATE_IP)) {
                 $ipAddress = $remoteAddr;
             }
         }
+        if (!$this->checkProxyHeaders) {
+            // do not check if configured to not check
+            return $ipAddress;
+        }
 
-        if ($this->shouldCheckProxyHeaders($ipAddress)) {
-            foreach ($this->headersToInspect as $header) {
-                if ($request->hasHeader($header)) {
-                    $ip = $this->getFirstIpAddressFromHeader($request, $header);
-                    if ($this->isValidIpAddress($ip)) {
-                        $ipAddress = $ip;
-                        break;
-                    }
+        // If trustedProxies is empty, then the remote address is the trusted proxy
+        $trustedProxies = $this->trustedProxies;
+        if (empty($trustedProxies) && empty($this->trustedWildcards) && empty($this->trustedCidrs)) {
+            $trustedProxies[] = $ipAddress;
+        }
+
+        // find the first non-empty header from the headersToInspect list and use just that one
+        foreach ($this->headersToInspect as $header) {
+            if ($request->hasHeader($header)) {
+                $headerValue = $request->getHeaderLine($header);
+                if (!empty($headerValue)) {
+                    $ipAddress = $this->getIpAddressFromHeader(
+                        $header,
+                        $headerValue,
+                        $ipAddress,
+                        $trustedProxies,
```

Comment:
This comma will cause issues in versions earlier than 7.3.

Reply:
Great catch! Thank you.